datafusion/expr/src/expr_schema.rs

```diff
@@ -128,7 +128,14 @@ impl ExprSchemable for Expr {
             Expr::Like { .. } | Expr::ILike { .. } | Expr::SimilarTo { .. } => {
                 Ok(DataType::Boolean)
             }
-            Expr::Placeholder { data_type, .. } => Ok(data_type.clone()),
+            Expr::Placeholder { data_type, .. } => {
+                let data_type = data_type.clone().ok_or_else(|| {
+                    DataFusionError::Internal(
+                        "Placeholder type could not be resolved".to_owned(),
+                    )
+                })?;
+                Ok(data_type)
```

Comment:
```suggestion
                data_type.clone().ok_or_else(|| {
                    DataFusionError::Internal(
                        "Placeholder type could not be resolved".to_owned(),
                    )
                })
```

Reply:
very minor style suggestion

Reply:
I also think this error might be better "Plan" or "Unsupported" -- it isn't really an internal (unexpected) error type